File: lib/Python/numbers.py
import math
import functools

def primes_up_to(n):
    primes = []
    i = 2
    while i <= n:
        if is_prime(i, primes):
            primes.append(i)
        i = i + 1
    return primes

def is_prime(i, primes):
    return not functools.reduce(
                lambda a, b: a or b, 
                [i % p == 0 for p in primes], 
                False)

class Factors:
    def __init__(self, n):
        self.n = n
        self.low_factors_cache = []
        self.high_factors_cache = []
        
    def factors(self):
        return self.low_factors() + self.high_factors()
        
    def low_factors(self):
        if self.low_factors_cache == []:
            self.low_factors_cache = list(filter(lambda a: self.n % a == 0, range(1, int(math.sqrt(self.n))+1)))
        return self.low_factors_cache
        
    def high_factors(self):
        if self.high_factors_cache == []:
            self.high_factors_cache = list(map(lambda a: self.n // a, reversed(self.low_factors())))
        return self.high_factors_cache

def factors(n):
    if n == 1:
        return [1]
    return Factors(n).factors()
    
def triangle_number(n):
    return sum(range(1, n+1))

File: lib/Erlang/numbers.erl
-module(numbers).
-export([primes_up_to/1, factorial/1, factors/1, triangle_number/1]).

-import(lists, [any/2, reverse/1, seq/2]).
-import(math, [sqrt/1]).

primes_up_to(N) -> reverse(primes_up_to(N, 2, [])).

primes_up_to(N, I, Primes) when I > N -> Primes;
primes_up_to(N, I, Primes) ->
  case is_prime(I, Primes) of
    true -> primes_up_to(N, I+1, [I | Primes]);
    false -> primes_up_to(N, I+1, Primes)
  end.
  
is_prime(I, Primes) -> not any(fun(P) -> I rem P == 0 end, Primes).

factors(1) -> [1];
factors(N) -> 
  IsDivisor = fun(F) -> N rem F == 0 end,
  OtherFactor = fun(F) -> N div F end,
  LowFactors = lists:filter(IsDivisor, seq(1, round(sqrt(N)))),
  HighFactors = lists:map(OtherFactor, LowFactors),
  lists:append(LowFactors, lists:reverse(HighFactors)).

factorial(N) -> lists:foldl(fun (I, Acc) -> I * Acc end, 1, seq(1, N)).

triangle_number(N) -> lists:sum(seq(1, N)).
